### DISCUSSION TITLE
Linq SelectMany yield return

### DISCUSSION THREAD

**Discussion by denisenko93** [dotnet/csharplang#4601] [Language Ideas]
Allow use of yield return in lambda functions.

Example:

```
public class Client
{
   public string Phone1 {get;set;}
   public string Phone2 {get;set;}
}
```
Does not works:

```
var phones = clients.SelectMany(x =>
{
   yield return x.Phone1;
   yield return x.Phone2;
});
```
Only methods, operators and accessors could contain 'yield' statement
    
Works:

```
var phones = clients.SelectMany(GetPhones);

IEnumerable<string> GetPhones(Client client)
{
   yield return client.Phone1;
   yield return client.Phone2;
}
```

**Comment by mairaw** [comment]
@jaredpar @MadsTorgersen can you help? Should this be transferred somewhere?

**Comment by jaredpar** [comment]
This can be transferred to dotnet/csharplang but it's going to be likely resolved as a duplicate / "by design" there. This was an explicit decision by the language design team and it's been re-confirmed several times in the recent past.

**Reply by huoyaoyuan** [reply]
Could you have more explanation here?

**Reply by bernd5** [reply]
Currently `yield return` is only allowed in regular methods and local-functions but not in lambdas.
I assume it was rejected due to the required complexity...

**Reply by jaredpar** [reply]
The issue isn't technical complexity. There are challenges but they're more about just making existing code work in more scenarios. The issue is more about priority and whether or not it's deemed appropriate that lambdas should do this. LDM has revisited this a few times and essentially said "not enough motivation to do it now". 

Unfortunately we didn't write that down in any notes I can find. Hence I can't remember if we decided "not enough priority" or "not appropriate". In the past there was some sentiment around "not appropriate" but not sure if that generally remains true.

**Reply by HaloFour** [reply]
There's this comment from back on Roslyn:

https://github.com/dotnet/roslyn/issues/24#issuecomment-70448545

**Comment by aetos382** [comment]
I think it would be useful to be able to stop pipeline by returning zero results from `SelectMany`.
For pre-defined types, I can implement this by returning `Array.Empty<T>()` from `SelectMany`. However, if the return type is an anonymous type, this is not possible.
It can be done by using a tuple instead of an anonymous type, but the following code is long and not readable.

```cs
return Array.Empty<(TypeOfX, TypeOfY, TypeOfZ)>();
```

I would prefer to write a simple `yield break` instead.
